pkg/operator/controller/controller_lb.go

```diff
@@ -41,20 +40,37 @@ func (r *reconciler) ensureLoadBalancerService(ci *operatorv1.IngressController,
 	if err != nil {
 		return nil, err
 	}
-	if desiredLBService != nil && currentLBService == nil {
+
+	desired, exists := desiredLBService != nil, currentLBService != nil
+
+	switch {
+	case desired && !exists:
 		if err := r.client.Create(context.TODO(), desiredLBService); err != nil {
 			return nil, fmt.Errorf("failed to create load balancer service %s/%s: %v", desiredLBService.Namespace, desiredLBService.Name, err)
 		}
 		log.Info("created load balancer service", "namespace", desiredLBService.Namespace, "name", desiredLBService.Name)
 		return desiredLBService, nil
+	case !desired && exists:
+		err := r.finalizeLoadBalancerService(ci, dnsConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to finalize load balancer service %s: %v", currentLBService.Name, err)
+		}
+		current, err := r.currentLoadBalancerService(ci)
```

Comment:
`finalizeLoadBalancerService()` will update the service resource under our noses; something's probably off with the factoring here, but that's why I fetched it again

Reply:
If I'm reading the code correctly, as you mention, `finalizeLoadBalancerService()` updates finalizers for the service. If the DNS pieces of `finalizeLoadBalancerService()` were broken out into a separate func that you call, I think would address the issue. That can be done in a future PR though.

Reply:
finalizeLoadBalancerService could also take the target `Service`, which I _think_ would be more consistent with other mutating functions we have

Reply:
What I'm missing here is how to marry this to the LB service finalization... would like to hop on a call to discuss

Reply:
One way would be to bite the bullet and start persisting resources to represent the DNS records, which would get rid of this tricky coupling between DNS and the service record. For example, ensureDNS could create and delete specially labelled configmaps w/ the target values (reminiscent of CRDs we might use w/ external-dns), and the DNS manager could be changed so that instead of being a library called directly, the DNS manager keeps DNS in sync w/ the configmaps. I'm starting to become afraid that unless we get DNS in line with the rest of our reconciliation model, then by introducing mutability around DNS outside the create/delete flow we're going to end up with something unintelligible and prone to breaking, especially given the lack of tests...